src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
+            {
+                builder.Add(CSharpDirective.Parse(sourceFile, line.SpanIncludingLineBreak, match.Groups[1].Value, match.Groups[2].Value));
+            }
+        }
 
-              <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
-
-              <!--
-                Override targets which don't work with project files that are not present on disk.
-                See https://github.com/NuGet/Home/issues/14148.
-              -->
-
-              <Target Name="_FilterRestoreGraphProjectInputItems"
-                      DependsOnTargets="_LoadRestoreGraphEntryPoints"
-                      Returns="@(FilteredRestoreGraphProjectInputItems)">
-                <ItemGroup>
-                  <FilteredRestoreGraphProjectInputItems Include="@(RestoreGraphProjectInputItems)" />
-                </ItemGroup>
-              </Target>
-
-              <Target Name="_GetAllRestoreProjectPathItems"
-                      DependsOnTargets="_FilterRestoreGraphProjectInputItems"
-                      Returns="@(_RestoreProjectPathItems)">
-                <ItemGroup>
-                  <_RestoreProjectPathItems Include="@(FilteredRestoreGraphProjectInputItems)" />
-                </ItemGroup>
-              </Target>
-
-              <Target Name="_GenerateRestoreGraph"
-                      DependsOnTargets="_FilterRestoreGraphProjectInputItems;_GetAllRestoreProjectPathItems;_GenerateRestoreGraphProjectEntry;_GenerateProjectRestoreGraph"
-                      Returns="@(_RestoreGraphEntry)">
-                <!-- Output from dependency _GenerateRestoreGraphProjectEntry and _GenerateProjectRestoreGraph -->
-              </Target>
-            </Project>
-            """;
-        ProjectRootElement projectRoot;
-        using (var xmlReader = XmlReader.Create(new StringReader(projectFileText)))
+        // The result should be ordered by source location, RemoveDirectivesFromFile depends on that.
+        return builder.ToImmutable();
+    }
+
+    public static SourceFile CreateSourceFile(string filePath)
+    {
+        using var stream = File.OpenRead(filePath);
+        return new SourceFile(filePath, SourceText.From(stream, Encoding.UTF8));
+    }
+
+    public static SourceText? RemoveDirectivesFromFile(ImmutableArray<CSharpDirective> directives, SourceText text)
+    {
+        if (directives.Length == 0)
+        {
+            return null;
+        }
+
+        Debug.Assert(directives.OrderBy(d => d.Span.Start).SequenceEqual(directives), "Directives should be ordered by source location.");
+
+        for (int i = directives.Length - 1; i >= 0; i--)
+        {
+            var directive = directives[i];
+            text = text.Replace(directive.Span, string.Empty);
+        }
+
+        return text;
+    }
+
+    public static void RemoveDirectivesFromFile(ImmutableArray<CSharpDirective> directives, SourceText text, string filePath)
+    {
+        if (RemoveDirectivesFromFile(directives, text) is { } modifiedText)
```

Comment:
In what case is this false? And should it mention to the caller that it failed?

Reply:
This is `false` if `RemoveDirectivesFromFile(directives, text)` returns `null` which happens if there are no directives, i.e., no changes to the file are necessary. The method could return `bool`, sure, but currently the caller doesn't need it. It is not a failure, it is just a no-op.